lib/phoenix_component.ex

```diff
@@ -1310,10 +1347,16 @@ defmodule Phoenix.Component do
     raise_bad_socket_or_assign!("changed?/2", assigns)
   end
 
+  defmacro embed_templates(root_dir \\ nil, pattern) do
+    root_dir = root_dir || Path.dirname(__CALLER__.file)
+    Module.put_attribute(__CALLER__.module, @phoenix_embeds, {root_dir, pattern})
```

Comment:
It is best to not do this at expansion time. I think we should also not delay the embedding, so we don't run into issues with delaying the expansion and things not working. I would roll with something like this:

```suggestion
  defmacro embed_templates(root_dir \\ nil, pattern) do
    quote do
        Phoenix.Template.compile_all(
          &Phoenix.Component.__embed__/1,
          unquote(root_dir) || __DIR__,
          unquote(pattern)
        )
    end
  end
```

(I am assuming we don't need to actually define `__templates__`).

Reply:
Actually, I would maybe go with another signature:

```suggestion
  defmacro embed_templates(pattern, opts \\ []) do
    quote do
        Phoenix.Template.compile_all(
          &Phoenix.Component.__embed__/1,
          unquote(opts)[:root] || __DIR__,
          unquote(pattern)
        )
    end
  end
```

It should be safer in the long term. :)